### ISSUE TITLE
bliss: please export

### ISSUE THREAD

**Issue by jgmbenoit** [igraph/igraph#1586]
I have noticed that the coming minor version 0.9 uses a slightly modified version of bliss. These changes are not mathematical ones, but they concern the interface. I would suggest to submit a patch to the upstream maintainer of bliss as your change may interest other developers. This will allow you to externalize bliss. This is better for everyone: as developer, you get ride of a external piece of  source; as package maintainer, you are certain to deal with only one version of bliss so that you can easily fix issues, to update, and so forth; as user, you are also certain that you deal with only one version.

**Comment by szhorvat** [comment]
Yes, I am in contact with the maintainer and this is currently in progress.

**Comment by jgmbenoit** [comment]
I can see that the bliss material in version 0.8.5 has been touched as well. So bliss cannot be exported for version 0.8.5 too.
I am considering to export cs stuff in 0.8.5: is it a good idea ?

**Comment by ntamas** [comment]
CS can probably be exported; we already support linking to an external CS in 0.9.0 with CMake and as I recall we did not need to modify anything to make it work.

T.

**Comment by jgmbenoit** [comment]
Thanks for the reply. Indeed the exporting is easy.

**Comment by szhorvat** [comment]
> Indeed the exporting is easy.

As Tamás said, this is probably fine for CXSparse. However, please do not do it for any of the other libraries, and definitely do not do it without asking. It is not as easy as it looks. For example, CXSparse uses the random number generator, which means that normally it would need to be modified to use igraph's own RNG (if you look in the sources, it *is* modified). It could be externalized only because the function that uses the RNG happens not to be called by igraph. These things must be carefully verified.

**Comment by jgmbenoit** [comment]
I suspected something like that. Hence my question.

**Comment by szhorvat** [comment]
Thinking about this, I am not in favour of trying to link to an external Bliss for the following reasons:

 1. It provides absolutely no benefits to users.
 2. It theoretically provides some benefits to us (developers) by making library upgrades easier, but this potential benefit is conditional on assumptions that might not hold, just as Bliss keeping a stable API (unlikely). At the moment, as the person dealing with Bliss within igraph, I do not feel that upgrading is difficult.
 3. It would take a considerable amount of work to make it happens (for very little benefit as mentioned above):
    - Bliss is not currently a reusable library, but a command line program. It would need to be changed into a proper library with a well-defined interface.
    - igraph integrates Bliss's error handling, which currently requires modifications (more work to expose in a proper library)
    - We needed to make modifications to access certain information such as the exact automorphism group size (more work to expose in a proper library).
 4. Bliss can be compiled with or without GMP, but igraph _requires_ the version with GMP. If it's a separate library, how do we guarantee it has the right configuration?
 5. It would become the first external library that is _required_, making out users' lives harder. Note that most users are researchers, not programmers. All other libraries igraph currently relies on are either vendorable, or are required only for optional (i.e. disableable) features. Bliss is essential to igraph and it is not acceptable to make it disableable.
 6. It would force GMP to be a required dependency as well (currently we can simply include Mini-GMP, which is at this moment perfectly fine as it's not performance-critical).
 7. CRAN rules unfortunately put constraints on igraph. Bliss is not available on CRAN, so we would either have to make it optional (not acceptable) or include it somehow. Right now we include it, which means that we need to make it CRAN compliant (no references to symbols such as `printf`, `stdout`, `abort`, etc.) Thus custom modifications to Bliss seem unavoidable.

In short, attempting to link to an external version would be a lot of work for no tangible gain.
